### PR TITLE
Run natural language tasks on a graph

### PR DESCRIPTION
**Pull request by darabos**
For #348. I went back and forth between a lot of different approaches.

- Use SQL. This was the original plan. We could augment the SQL box to use a new backend call to generate SQL. Not a bad option. It's great that the user would see the SQL and be able to tweak it. But we would have to do the OpenAI call from Scala, and it would take some frontend work. (A new "param".) Not exactly a quick prototype. Plus it wouldn't work well with custom boxes.
- Use SQL in Python. Adding some code to a Python box is much simpler for a quick prototype. We could use `dfsql` to run a SQL query like in https://github.com/lynxkite/lynxkite/pull/349.
- Generate Pandas code instead of SQL. Codex has probably seen more Pandas code than SQL.
- In either case, what's a good example query? The Python box can only compute new attributes. It can't find the most friendly chef or list the cities connected to Cincinnati. I'm going to do #352 to address this and allow outputting a table based on a graph.
- Another difficulty with a Python box is that it has to specify the output schema. I had some wild ideas, like doing the Codex request in a parametric parameter. But in the end @lacca0 made a good argument for letting the user specify the schema by hand. Then we put it in the prompt. Con: Codex may ignore it and build the results with a different schema. Pro: specifying the schema clarifies the user's intent.

In the end I'm going down the Python box route now. Even if we add the new endpoint later this looks quite useful.

### CODE REVIEW

File: sphynx/python/llm_pandas_on_graph.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+import sys
+import openai
+# Langchain is not on Conda yet. Run "pip install langchain".
+# https://github.com/hwchase17/langchain/issues/1271
+import langchain
+
+openai.api_key = os.getenv('OPENAI_API_KEY')
+
+
+examples = [
```
Comment (darabos) [inline sphynx/python/llm_pandas_on_graph.py:14]: @lacca0 This is the kind of thing I tried to describe in #348. We have some scenarios and we can automatically put them in a prompt (`function_def`) that we can tweak separately. The examples seem very important to me. Without them Codex generated me stuff like:
```python
  # YOUR CODE HERE
  #raise NotImplementedError()
```

**Comment by darabos** [comment]
It's working!

image

Unfortunately, it also shows reliability issues with this. On the first try it generated this code:
```python
def compute_from_graph(nodes, edges):
  """
  Solves the task: compute the average income of their neighbors

  "nodes" is a Pandas DataFrame with the following columns:
      age  gender id  income                    location          name
0  20.3    Male  0  1000.0       [40.71448, -74.00598]          Adam
1  18.2  Female  1     NaN    [47.5269674, 19.0323968]           Eve
2  50.3    Male  2  2000.0      [1.352083, 103.819836]           Bob
3   2.0    Male  3     NaN  [-33.8674869, 151.2069902]  Isolated Joe

  "edges" is a Pandas DataFrame with the following columns:
             comment  weight  src  dst
0   Adam loves Eve     1.0    0    1
1   Eve loves Adam     2.0    1    0
2  Bob envies Adam     3.0    2    0
3    Bob loves Eve     4.0    2    1

  Returns a DataFrame with columns: name: str, avg_neighbor_income: float
  """
  # compute the average income of their neighbors
```
```python
  edges = edges.merge(nodes.add_suffix('_src'), left_on='src', right_index=True)
  edges = edges.merge(nodes.add_suffix('_dst'), left_on='dst', right_index=True)
  # Keep only edges where the source and destination are different.
  edges = edges[edges['id_src'] != edges['id_dst']]
  nodes['avg_neighbor_income'] = edges.groupby('id_src')['income_dst'].mean()
  return nodes
```

The bit before the break is part of the prompt. I include it to show that there is no mention of `id_src` and `id_dst`. Codex made that up!

The second try you see on the screenshot generated this code:

```python
  edges = edges.merge(nodes.add_suffix('_src'), left_on='src', right_index=True)
  edges = edges.merge(nodes.add_suffix('_dst'), left_on='dst', right_index=True)
  # Keep only edges where both nodes have income.
  edges = edges[edges['income_src'].notnull() & edges['income_dst'].notnull()]
  nodes['avg_neighbor_income'] = edges.groupby('src')['income_dst'].mean()
  return nodes
```

Fairly good! I love that it added a comment for the `edges where both nodes have income` logic. That restriction may or may not be what the user intended. This sort of ambiguity is not exclusive to natural language. When writing code, we would be likely to err on the side of not adding such a filter. Then we get whatever happens by default. That may or may not be what we intended. Such semantics can be hard to debug in code too, but at least I have decades of experience with that.

Anyway, let me package this up in a built-in custom box!

**Comment by darabos** [comment]
The trick is to ask simple questions!

image

Perfect! ⭐ ⭐ ⭐ ⭐ ⭐

**Comment by darabos** [comment]
I've become a fan of the user-specified schema. This makes it possible to expose the query parameter in a wizard, but apply follow-up steps to the result table. Check it out in action:

zh_2200_
zh_2200_ (1)

zh_2200_ (3)
zh_2200_ (2)

zh_2200_ (5)
zh_2200_ (4)

The backing workspace:
image

"OpenAI-airlines" is just the OpenAI box with the output schema fixed to `airport: str`.

**Comment by darabos** [comment]
The test failures are due to some Yarn shenanigans:

```
 --> RUN cd web && yarn --frozen-lockfile
           +npm-deps *failed* | ➤ YN0050: The --frozen-lockfile option is deprecated; use --immutable and/or --immutable-cache instead
           +npm-deps *failed* | ➤ YN0070: Migrating from Yarn 1; automatically enabling the compatibility node-modules linker 👍

           +npm-deps *failed* | ➤ YN0000: ┌ Resolution step
           +npm-deps *failed* | ➤ YN0061: │ asciidoctor.js@npm:1.5.9 is deprecated: Package no longer supported. Replaced by @asciidoctor/core
           +npm-deps *failed* | ➤ YN0061: │ angular-route@npm:1.8.2 is deprecated: For the actively supported Angular, see https://www.npmjs.com/package/@angular/core. AngularJS support has officially ended. For extended AngularJS support options, see https://goo.gle/angularjs-path-forward.
           +npm-deps *failed* | ➤ YN0061: │ angular-sanitize@npm:1.8.2 is deprecated: For the actively supported Angular, see https://www.npmjs.com/package/@angular/core. AngularJS support has officially ended. For extended AngularJS support options, see https://goo.gle/angularjs-path-forward.
           +npm-deps *failed* | ➤ YN0061: │ angular-cookies@npm:1.8.2 is deprecated: For the actively supported Angular, see https://www.npmjs.com/package/@angular/core. AngularJS support has officially ended. For extended AngularJS support options, see https://goo.gle/angularjs-path-forward.
           +npm-deps *failed* | ➤ YN0061: │ protractor@npm:7.0.0 is deprecated: We have news to share - Protractor is deprecated and will reach end-of-life by Summer 2023. To learn more and find out about other options please refer to this post on the Angular blog. Thank you for using and contributing to Protractor. https://goo.gle/state-of-e2e-in-angular
           +npm-deps *failed* | ➤ YN0061: │ angular@npm:1.8.2 is deprecated: For the actively supported Angular, see https://www.npmjs.com/package/@angular/core. AngularJS support has officially ended. For extended AngularJS support options, see https://goo.gle/angularjs-path-forward.
           +npm-deps *failed* | ➤ YN0061: │ request@npm:2.88.2 is deprecated: request has been deprecated, see https://github.com/request/request/issues/3142
           +npm-deps *failed* | ➤ YN0061: │ har-validator@npm:5.1.5 is deprecated: this library is no longer supported
           +npm-deps *failed* | ➤ YN0061: │ source-map-resolve@npm:0.6.0 is deprecated: See https://github.com/lydell/source-map-resolve#deprecated
           +npm-deps *failed* | ➤ YN0061: │ uuid@npm:3.4.0 is deprecated: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
           +npm-deps *failed* | ➤ YN0061: │ chokidar@npm:2.1.8 is deprecated: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies
           +npm-deps *failed* | ➤ YN0061: │ source-map-resolve@npm:0.5.3 is deprecated: See https://github.com/lydell/source-map-resolve#deprecated
           +npm-deps *failed* | ➤ YN0061: │ urix@npm:0.1.0 is deprecated: Please see https://github.com/lydell/urix#deprecated
           +npm-deps *failed* | ➤ YN0061: │ resolve-url@npm:0.2.1 is deprecated: https://github.com/lydell/resolve-url#deprecated
           +npm-deps *failed* | ➤ YN0061: │ source-map-url@npm:0.4.1 is deprecated: See https://github.com/lydell/source-map-url#deprecated
           +npm-deps *failed* | ➤ YN0061: │ debug@npm:4.1.1 is deprecated: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
           +npm-deps *failed* | ➤ YN0032: │ fsevents@npm:2.3.2: Implicit dependencies on node-gyp are discouraged
           +npm-deps *failed* | ➤ YN0061: │ fsevents@npm:1.2.13 is deprecated: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
           +npm-deps *failed* | ➤ YN0032: │ nan@npm:2.14.2: Implicit dependencies on node-gyp are discouraged
           +npm-deps *failed* | ➤ YN0061: │ @npmcli/move-file@npm:2.0.1 is deprecated: This functionality has been moved to @npmcli/fs
           +npm-deps *failed* | ➤ YN0061: │ @types/fast-json-stable-stringify@npm:2.1.0 is deprecated: This is a stub types definition. fast-json-stable-stringify provides its own type definitions, so you do not need this installed.
           +npm-deps *failed* | ➤ YN0061: │ core-js@npm:3.6.5 is deprecated: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.
           +npm-deps *failed* | ➤ YN0002: │ @firebase/auth@npm:0.16.6 [965bc] doesn't provide @firebase/app-types (pa30b6), requested by @firebase/auth-types
           +npm-deps *failed* | ➤ YN0002: │ @firebase/auth@npm:0.16.6 [965bc] doesn't provide @firebase/util (p0cd1e), requested by @firebase/auth-types
           +npm-deps *failed* | ➤ YN0002: │ @firebase/database@npm:0.10.4 doesn't provide @firebase/app-types (pb2b5c), requested by @firebase/auth-interop-types
           +npm-deps *failed* | ➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
           +npm-deps *failed* | ➤ YN0000: └ Completed in 23s 187ms
           +npm-deps *failed* | ➤ YN0000: ┌ Post-resolution validation
           +npm-deps *failed* | ➤ YN0028: │ The lockfile would have been modified by this install, which is explicitly forbidden.
           +npm-deps *failed* | ➤ YN0000: └ Completed in 0s 216ms
```

I don't think the deprecations are blocking the build, but the `doesn't provide` lines may. Anyway, I'll try to sort this out in another PR. Please go ahead with the review in the meanwhile! Thanks!

**Review by lacca0 — APPROVED** [review]
Thanks for all the work. This seem to be really convenient! Let's empasize this possibility for the users somehow. Maybe a quick blog post?

**Comment by darabos** [comment]
Okay. We have some more ideas for this before the release, but let's do that in another PR. Thanks for the review!

**Pull request merged**